app/controllers/concerns/force_data_use_terms_agreement.rb

```
@@ -0,0 +1,42 @@
+module ForceDataUseTermsAgreement
+  extend ActiveSupport::Concern
+
+  included do
+    before_action :require_data_use_terms
+  end
+
+  private
+  def require_data_use_terms
+    return if current_scope == "admin" || current_scope == "regional_ambassador"
+
+    if logged_in_and_has_profile && no_data_agreement && not_on_data_agreement_form
+      redirect_to account_data_data_use_terms_edit_path,
+        notice: t("controllers.application.no_data_agreement_on_file") and return
+    end
+  end
+
+  def logged_in_and_has_profile
+    current_account.authenticated? &&
+      current_account.respond_to?("#{current_scope}_profile") &&
+        !!current_account.send("#{current_scope}_profile")
+  end
+
+  def no_data_agreement
+    !current_account.terms_agreed?
+  end
+
+  def not_on_data_agreement_form
+    return false if request.xhr?
+
+    begin
```

What's the scenario where Rails wouldn't be able to recognize the request path or the data use terms edit path? Is it if the user puts in a garbage url or something?

If Rails can't recognize `account_data_data_use_terms_edit_path` then something is very wrong, so maybe that should be outside of the begin block to fail fast.

---

It failed for me during a redirect. For some reason it was trying to visit the "/judge/location" path without passing the right parameters, so I think it might need reassessment. I will dive into this a bit more.